### PR TITLE
feat(ui): replace navigation with test explorer

### PR DESCRIPTION
**Pull request by userquin**
### Description

This PR changes the navigation to match the test explorer in VSCode extension:
- remove all vue reactivity from ws client files and file ids, only reactivity in state is required: the server will not handle backpressure on the client and so the current version will hang when there are a lot of files and tests (tested with `test/core`)
- use only 1 detail panel in the navigation with a virtual scroll: included `useVirtualScroll` from VueUse, **we need to use another vue library since the tasks will have children (suite/tests...)** (with children expanded the VueUse virtual scroll will not handle the height properly)
- load files on ws open event, sorting the files received in the initial call, this way we don't need to sort them, sort will be done only once
- added `setInterval` to collect the summary state in `composables/test-state.ts` (this PR will not use `composables/summary.ts`): we need to adjust the timeout, we cannot use `setTimeout`  approach, the server will not handle backpressure and the ui can miss a lot of ticks (check `_collect` function in `composables/test-state.ts`)
- logic to load file children using just the `id` and reload the task when the id changes
- allow `@vitest/ws-client` to change the reactivity based on the data (check `VitestClientOptions` and `createClient`), the current version will use `reactive` for all, that's, files, file ids and state
- filter by status
- TODO: use vue virtual scroller

Maybe we need to move the logic to collect global state to the server, this way the client will just receive the data to be displayed in the ui.

<details>
<summary><i>Second run: the server finished and the client still processing incoming data</i></summary>

imagen
</details>

<details>
<summary><i>Second run: the CPU 100%, once server finish only 1 core (without dev tools)</i></summary>

imagen
</details>

<details>
<summary><i>Vitest Browser tests (preview provider)</i></summary>

imagen
</details>

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: packages/ui/client/composables/client/index.ts [hunk 1]
```diff
@@ -2,33 +2,79 @@ import { createClient, getTasks } from '@vitest/ws-client'
...
   else {
+    // const callbacks = /* #__PURE__ */ new WeakMap<{ key: string }, () => void>()
+    function onCollected(collectedFiles: File[] = [], cb?: () => void) {
+      try {
+        // trigger update to all files
+        collectedFiles.forEach((file) => {
+          const entry = files.value.find(f => f.id === file.id)
+          if (entry) {
+            // callbacks.get({ key: file.id })?.()
+            /* const number = */requestAnimationFrame(() => {
```
Comment (JessicaSachs) [inline packages/ui/client/composables/client/index.ts:35]: Maybe I'm missing some cleanup, but you're not properly tearing down the `requestAnimationFrame` anymore, yeah? Shouldn't we be using `useRequestAnimationFrame` or something from VueUse to ensure we're not leaking?

**Review by JessicaSachs — COMMENTED** [review]
I might be missing something wrt the `requestAnimationFrame` loop cleanup. Looks like there's a leak.

**Comment by userquin** [comment]
Latest changes:
- updated the components and removed `New**` components (ListView should be also removed)
- `test-state.ts` moved to `summary.ts`
- right now we need to manually start the RAF `collect` loop, I'll try to start/stop when required

There are 2 todos in the code @sheremet-va  :
- since we need to start manually in the code the RAF loop, we need to check if the server is still running, for example a page refresh once finished or while running: check if the code here is fine [checkFinished](https://github.com/vitest-dev/vitest/pull/5854/files#diff-ccf62853bfdf7af25e3724a99331864caf9f10999cb4bb5d1c46156d72b808f3R160-R164)
- when running some test (one file/test/suite) or just filtering and running, the dashboard will show all the files, whereas in the server we only have the test we run: we can switch the behavior to match the server output (we also need to check browser, but should be fine) => [clearResults](https://github.com/vitest-dev/vitest/pull/5854/files#diff-ca4e0265ae5d632ea5bfaeb4a603fe0d5e0992ac248197ae441d867e0bcb3b04R75-R77)

I'm going to include `vue-virtual-scroller` with recycler sroller and [variable size mode](https://github.com/Akryum/vue-virtual-scroller/blob/master/packages/vue-virtual-scroller/README.md#variable-size-mode) (iirc @patak-dev  did a great work at elk.zone)

**Comment by userquin** [comment]
Included summary in details panel, the ui tests checking for old layout where we have 4 panels, now we have only 1.

imagen

File: packages/browser/src/client/orchestrator.ts [hunk 2]
```diff
@@ -50,9 +50,14 @@ function createIframe(container: HTMLDivElement, file: string) {
...
 
-async function done() {
-  await rpcDone()
-  await client.rpc.finishBrowserTests()
+async function done(testFinished = true) {
+  try {
+    await rpcDone()
+    await client.rpc.finishBrowserTests()
+  }
+  finally {
```
Comment (sheremet-va) [inline packages/browser/src/client/orchestrator.ts:58]: Why is this needed? Finishing should be done in `onFinished` hook in UI

Comment (userquin) [reply packages/browser/src/client/orchestrator.ts:58]: not being called

Comment (userquin) [reply packages/browser/src/client/orchestrator.ts:58]: Maybe being called without files and with errors, I'm going to check what's happening, maybe the logic must be updated for browser mode.

Comment (sheremet-va) [reply packages/browser/src/client/orchestrator.ts:58]: Calling `finish` resolves this promise:

https://github.com/vitest-dev/vitest/blob/07876b7e0f4f69cd071f0c89107ebdd08927e182/packages/vitest/src/node/pools/browser.ts#L59

Which in turn resolves this:

https://github.com/vitest-dev/vitest/blob/07876b7e0f4f69cd071f0c89107ebdd08927e182/packages/vitest/src/node/core.ts#L636

Which should always call `onFinished` here:

https://github.com/vitest-dev/vitest/blob/07876b7e0f4f69cd071f0c89107ebdd08927e182/packages/vitest/src/node/core.ts#L655

Comment (userquin) [reply packages/browser/src/client/orchestrator.ts:58]: `done` being called 3 times (I added a breakpoint)... will `onFinished` being called only once, the first time it is resolved? Then the client can receive some `onTaskUpdate` and the state will be reset to `running` again

Comment (userquin) [reply packages/browser/src/client/orchestrator.ts:58]: `onFinished` being called only once but we're receiving `onTaskUpdate` after `onFinished` called:

imagen

**Comment by userquin** [comment]
Now tasks also being filtered:

imagen

imagen

File: packages/browser/src/client/orchestrator.ts (hunk 2 above)
Comment (userquin) [reply packages/browser/src/client/orchestrator.ts:58]: no idea what's happening but now works again without those changes: I'm going to revert changes

Comment (sheremet-va) [reply packages/browser/src/client/orchestrator.ts:58]: Looks like it is a bug in Vitest runner - it should wait until all `onTaskUpdate` are resolved before finishing `runTests`

**Comment by userquin** [comment]
I'm going to submit a new PR with latest changes in main, you can check [this video](https://streamable.com/sk17ed).

These are the pending stuff (can be progressive):
- store filter in session to restore it on page refresh
- add `openedTreeItems`  logic, right now removed in `Explorer.vue`
- add sintax hightlight to the tasks when searching
- allow run tests not just the file (will require some changes in the runner and ws-client)
- show all subtasks when`Only Tests`  is not checked and the files/suites match the search and there are no tests matching the search

To run some ui tests:
1) fork the repo, uncheck `Copy the main branch only`, clone to your local and checkout `[userquin/feat-add-test-explorer-ui](https://github.com/vitest-dev/vitest/tree/userquin/feat-add-test-explorer-ui)`  branch
2 ) build vitest, from root run `pnpm install --frozen-lockfile`
3 ) to run `test/core`, from root folder run `cd test/core && pnpm vitest --ui --open`
4) to run `test/browser`, from root folder run `cd test/browser && nr test:browser:preview`

File: packages/ui/client/composables/explorer/tree.ts [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+            const parent = this.nodes.get(child.parentId)
+            if (parent && 'filepath' in parent)
+              filesToShow.add(parent.id)
+
+            yield child
+          }
+        }
+        else {
+          // show the parent if matches the filter or at least one child matches the filter
+          if (match || treeNodes.has(child.id) || filesToShow.has(child.id)) {
```
Comment (userquin) [inline packages/ui/client/composables/explorer/tree.ts:483]: should check `filesToShow` only for files

File: packages/ui/client/composables/explorer/tree.ts [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+
+export { taskTree }
+
+declare global {
+  interface Window {
+    taskTree: TaskTree
+  }
+}
+
+window.taskTree = taskTree
```
Comment (userquin) [inline packages/ui/client/composables/explorer/tree.ts:594]: remove global and taskTree from window

File: packages/ui/client/composables/search.ts [hunk 5]
```diff
@@ -16,7 +19,7 @@ const filter = reactive(<Filter>{
   success: false,
   skipped: false,
 })
-// don't remove this computed, reactive filter is not working inside computed filtered
+// don't remove these computed, reactive filter is not working inside computed filtered
```
Comment (userquin) [inline packages/ui/client/composables/search.ts:22]: remove this

Comment (userquin) [reply packages/ui/client/composables/search.ts:22]: upps, this module will be removed

File: packages/ui/client/composables/explorer/search.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  failedFilter,
+  filter,
+  filteredFiles,
+  isFiltered,
+  isFilteredByStatus,
+  search,
+  skipFilter,
+  successFilter,
+} from './state'
+import type { TreeTaskFilter } from '~/composables/explorer/tree'
```
Comment (userquin) [inline packages/ui/client/composables/explorer/search.ts:14]: move all types to `types.ts`

File: packages/ui/client/composables/explorer/search.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    isFilteredByStatus,
+    disableClearSearch,
+    clearSearch,
+    clearFilter,
+    filesTotal: readonly(filesTotal),
+    uiEntries,
+  }
+}
+
+export function recalculateUITreeExplorer() {
```
Comment (userquin) [inline packages/ui/client/composables/explorer/search.ts:113]: remove this

File: packages/ui/client/composables/explorer/search.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+      }, { failed: 0, success: 0, skipped: 0, running: 0 })
+      data.running = filteredFiles.value.length - data.failed - data.success - data.skipped
+      return data
+    }
+
+    return {
+      failed: taskTree.summary.filesFailed,
+      success: taskTree.summary.filesSuccess,
+      skipped: taskTree.summary.filesSkipped,
+      running: taskTree.summary.filesRunning,
```
Comment (userquin) [inline packages/ui/client/composables/explorer/search.ts:50]: review logic, should be test and no files: move the logic to the tree to collect the state and rename `filteredFiles` back to `filteredTests`

File: packages/ui/client/composables/explorer/tree.ts [hunk 9]
```diff
@@ -0,0 +1,594 @@
...
+    this.summary.filesTodo = data.filesTodo
+    this.summary.testsFailed = data.testsFailed
+    this.summary.testsSuccess = data.testsSuccess
+    this.summary.testsFailed = data.testsFailed
+    this.summary.testsTodo = data.testsTodo
+    this.summary.testsIgnore = data.testsIgnore
+    this.summary.testsSkipped = data.testsSkipped
+    this.summary.totalTests = data.totalTests
+    // update filtered files
+    filteredFiles.value = this.filteredFiles()
```
Comment (userquin) [inline packages/ui/client/composables/explorer/tree.ts:283]: try to add new tasks here to the tree: if we have a filter applied while running, all missing subtasks will be loaded once `onFinished` called

File: packages/ui/client/composables/explorer/search.ts [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    filter.failed,
+    filter.success,
+    filter.skipped,
+    filter.onlyTests,
+    allExpanded.value,
+  ], ([search, failed, success, skipped, onlyTests, expandAllFlag]) => {
+    clearTimeout(taskId.value)
+    taskId.value = taskTree.buildNavigationEntries(
+      expandAllFlag,
+      search || failed || skipped || success ? (onlyTests ? showOnlyTests : dontShowOnlyTests) : undefined,
```
Comment (userquin) [inline packages/ui/client/composables/explorer/search.ts:95]: use only 1 `showOnlyTests`, just update the value

File: packages/ui/client/components/Navigation.vue [hunk 11]
```diff
@@ -40,38 +39,43 @@ async function onRunAll(files?: File[]) {
...
-  openedTreeItems.value = []
+  allExpanded.value = false
+  // openedTreeItems.value = []
 }
 
 function expandTests() {
-  files.value.forEach(file => {
+  allExpanded.value = true
+  /*files.value.forEach(file => {
     if (!openedTreeItems.value.includes(file.id)) {
```
Comment (userquin) [inline packages/ui/client/components/Navigation.vue:56]: add `openedTreeItems` logic again

File: packages/ui/client/composables/explorer/search.ts [hunk 12]
```diff
@@ -49,30 +44,57 @@ export function useSearch(searchBox: Ref<HTMLDivElement | undefined>) {
...
+      filter.success,
+      filter.skipped,
+      filter.onlyTests,
+    ] as const,
+    ([search, failed, success, skipped, onlyTests]) => {
+      // clearTimeout(taskId.value)
+      // defaultShowOnlyTests.showOnlyTests = onlyTests
+      updateFilterStorage(search, failed, success, skipped, onlyTests)
+      explorerTree.filterNodes()
+      // taskId.value = explorerTree.buildNavigationEntries(
```
Comment (userquin) [inline packages/ui/client/composables/explorer/search.ts:84]: cleanup

File: packages/ui/client/composables/explorer/state.ts [hunk 13]
```diff
@@ -29,16 +45,30 @@ export const isFilteredByStatus = computed(() => {
...
+  // - no openedTreeItems and expandAll = true
+  // - or expandAll = false (openedTreeItems can be empty or have entries, filteredFiles doesn't matter here)
+  // - or no openedTreeItems and expandAll = false (filtered items)
+  // - or openedTreeItems and expandAll = undefined (filtered items) **remove**
+
+  // const noItems = openedTreeItems.value.length === 0
+  return treeFilter.value.expandAll !== false
+
+  // return (noItems && expandAll === false) || expandAll === false/* || (!noItems && expandAll === undefined) */
+})
```
Comment (userquin) [inline packages/ui/client/composables/explorer/state.ts:61]: cleanup

File: packages/ui/client/composables/explorer/filter.ts [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+    nodes,
+    filter.onlyTests,
+    treeNodes,
+    filesToShow,
+  )].reverse()
+
+  // if (nodesToExpand.size === 0 && treeFilter.value.expandAll !== false) {
+  //   yield * entries
+  //   return
+  // }
```
Comment (userquin) [inline packages/ui/client/composables/explorer/filter.ts:79]: clenup

**Comment by userquin** [comment]
closing, superseded by  #5907
